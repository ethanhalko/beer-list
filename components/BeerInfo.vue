<script setup>
const props = defineProps({
  beer: {
    type: Object,
    required: true
  }
});

const readMore = shallowRef(false);
const containsLactose = !!props.beer.ingredients.hops.find((hops) => hops.name.toLowerCase() === 'lactose');
const dryHopped = !!props.beer.ingredients.hops.find((hops) => hops.add?.toLowerCase() === 'dry hop');
</script>
<template>
  <div class="beer-card group flex flex-col" :class="{'max-h-64 sm:max-h-36': !readMore}" @click="readMore = !readMore">
    <div class="flex flex-row">
      <div class="card-image w-[12.5%] sm:my-auto">
        <img class="mx-auto max-h-24" :class="{'sm:max-h-36': readMore}" :src="beer.image_url" :alt="beer.description">
      </div>
      <div class="card-info w-[87.5%] flex flex-col">
        <div class="flex">
          <h1 class="beer-name sm:text-xl px-1 -ml-1" :class="{'bg-green-200': dryHopped}">{{ beer.name }}</h1>
          <Icon v-if="containsLactose" name="fa6-solid:cow" class="icon-lactose my-auto text-middle mx-1 w-6 text-red-600"></Icon>
        </div>
        <h2 class="text-md sm:my-auto">ABV: {{ beer.abv }} | IBU: {{ beer.ibu }}</h2>
        <h3 class="sm:text-lg italic">{{ beer.tagline }}</h3>
        <p v-if="readMore" class="description sm:text-lg pt-4 mt-auto">{{ beer.description }}</p>
        <div class="w-[87.5%] mr-auto">
          <p v-if="!readMore" class="text-sm italic mx-auto text-center invisible group-hover:visible text-gray-400">Click To Expand</p>
        </div>
      </div>
    </div>
  </div>
</template>
<style>
.beer-card {
  @apply my-4 p-4 rounded-md border hover:shadow-md transition-shadow duration-200 cursor-pointer;
}
</style>

