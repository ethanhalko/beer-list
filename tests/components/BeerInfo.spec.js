import { describe, it, expect } from 'vitest';
import { mountSuspended } from 'nuxt-vitest/utils';
import BeerInfo from '../../components/BeerInfo.vue';

const testBeer = {
  id: 1,
  name: 'Saison D\'Ethan',
  description: 'A sour, hopped saison with notes of bergamot, cucumber',
  tagline: 'Great beer for writing tests',
  abv: 7,
  ibu: 65,
  image_url: '',
  ingredients: {
    hops: [
      {
        name: 'Cascade',
        amount: {
          value: 2.5,
          unit: 'grams'
        },
        add: 'start',
        attribute: 'bitter'
      },
      {
        name: 'Mosaic',
        amount: {
          value: 15,
          unit: 'grams'
        },
        add: 'end',
        attribute: 'flavour'
      },
      {
        name: 'Simcoe',
        amount: {
          value: 75,
          unit: 'grams'
        },
        add: 'dry hop',
        attribute: 'aroma'
      },
      {
        name: 'Lactose',
        amount: {
          value: 100,
          unit: 'grams'
        },
        add: 'middle',
        attribute: 'flavour'
      }
    ]
  }
};

describe('BeerInfoCard', async() => {
  const wrapper = await mountSuspended(BeerInfo, {
    global: {
      stubs: {
        fas: {
          template: '<span />',
        }
      }
    },
    props: { beer: testBeer }
  });

  it('renders a vue component', () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('highlights if dry hopped', () => {
    const name = wrapper.find('.beer-name');
    expect(name.exists()).toBeTruthy();
    expect(name.classes()).toContain('bg-green-200');
  });

  it('warns of lactose', () => {
    const lactoseIcon = wrapper.find('.icon-lactose');
    expect(lactoseIcon.exists()).toBeTruthy();
  });

  it('expands when clicked', async () => {
    await wrapper.trigger('click');
    expect(wrapper.find('.description').exists()).toBeTruthy();

    await wrapper.trigger('click');
    expect(wrapper.find('.description').exists()).toBeFalsy();
  });
});
