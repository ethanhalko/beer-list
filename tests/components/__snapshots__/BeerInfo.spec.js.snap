// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BeerInfoCard > matches snapshot 1`] = `
"<div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
  <div class=\\"flex flex-row\\">
    <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"\\" alt=\\"A sour, hopped saison with notes of bergamot, cucumber\\"></div>
    <div class=\\"card-info w-5/6 flex flex-col\\">
      <div class=\\"flex\\">
        <h1 class=\\"beer-name sm:text-xl px-1 -ml-1 bg-green-200\\">Saison D'Ethan</h1><span title=\\"Contains Lactose\\" class=\\"icon-lactose my-auto mx-2 text-red-600\\" icon=\\"fas,cow\\"></span>
      </div>
      <h2 class=\\"text-md sm:my-auto\\">ABV: 7 | IBU: 65</h2>
      <h3 class=\\"sm:text-lg italic\\">Great beer for writing tests</h3>
      <!--v-if-->
    </div>
  </div>
  <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
</div>"
`;
